src/uptane/tufrepository.cc

```diff
@@ -38,207 +49,95 @@ Json::Value TufRepository::getJSON(const std::string& role) {
-        if (keys_.count((*it_keyid).asString())) {
-          std::string key_can = Json::FastWriter().write(Json::Value(keys_[(*it_keyid).asString()].value));
-          std::string key_id = boost::algorithm::to_lower_copy(boost::algorithm::hex(Crypto::sha256digest(key_can)));
-          if ((*it_keyid).asString() != key_id) {
-            throw UnmetThreshold(name_, role);
-          }
-        }
-      }
-    }
-    thresholds_[role] += requiredThreshold;
-  }
-  verifyRole(content);
-}
-
+// See UPTANE Implementation Specification section 8.3.2
 void TufRepository::refresh() {
-  targets_.clear();
+  targets_.clear();  // TODO, this is used to signal 'no new updates'
+  updateRoot(Version());
   if (checkTimestamp()) {
-    Json::Value content = updateRole("snapshot.json");
-    Json::Value updated_roles = content["signed"]["meta"];
-    std::vector<std::string> files_to_update;
-    bool init_root = false;
-    for (Json::ValueIterator it = updated_roles.begin(); it != updated_roles.end(); it++) {
-      if (boost::ends_with(it.key().asString(), "root.json")) {
-        init_root = true;
-      } else {
-        files_to_update.push_back(it.key().asString());
-      }
-    }
-    if (init_root) {  // Root should be updated first
-      initRoot(updateRole("root.json"));
-    }
-    for (std::vector<std::string>::iterator it = files_to_update.begin(); it != files_to_update.end(); it++) {
-      Json::Value new_content = updateRole(*it);
-      if ((*it) == "targets.json") {
-        Json::Value json_targets = new_content["signed"]["targets"];
-        for (Json::ValueIterator t_it = json_targets.begin(); t_it != json_targets.end(); t_it++) {
-          Json::Value hashes = (*t_it)["hashes"];
-          Hasher hash;
-          if (hashes.isMember("sha512")) {
-            hash = Hasher(Hasher::sha512, hashes["sha512"].asString());
-          } else if (hashes.isMember("sha256")) {
-            hash = Hasher(Hasher::sha256, hashes["sha256"].asString());
-          }
-          Target t(Target((*t_it)["custom"], t_it.key().asString(), (*t_it)["length"].asUInt64(), hash));
-          saveTarget(t);
-        }
+    Json::Value snapshot_json = fetchAndCheckRole(kSnapshot);
+    // TODO snapshots
+
+    Json::Value targets_json = fetchAndCheckRole(kTargets);
+    Json::Value target_list = targets_json["targets"];
+    for (Json::ValueIterator t_it = target_list.begin(); t_it != target_list.end(); t_it++) {
+      Json::Value hashes = (*t_it)["hashes"];
+      int64_t length = (*t_it)["length"].asInt64();
+
+      Hash::Type type;
+      std::string hash_string;
+      if (hashes.isMember("sha512")) {
+        type = Hash::sha512;
+        hash_string = hashes["sha512"].asString();
+      } else if (hashes.isMember("sha256")) {
+        type = Hash::sha256;
+        hash_string = hashes["sha256"].asString();
       }
+      Hash hash(type, hash_string);
```

Comment:
We could reimplement Hash to accept `type` in string format, then we could iterate through `hashes` and append them to the list, then pass this list to the Target.